signac/contrib/utility.py

```diff
@@ -265,6 +406,18 @@ def _to_hashable(l):
 
 
 def _encode_tree(x):
+    """Encode if type of x is list.
+
+    Parameters
+    ----------
+    x :
+        type to encode.
```

Comment:
what should be the type of x here? and what exactly does encoding tree means here?

Reply:
I think it would be a list.

My best guess is that "encoding tree" is a reference ot the structure of nested statepoints, like this:
![image](https://user-images.githubusercontent.com/47036428/80753361-22978080-8afb-11ea-8d30-833f24238e62.png)

Reply:
but if you see there is a check in the function `_encode_tree` that returns x as it is if its type is not list.

Reply:
@klywang is correct about how this works. You're right that it returns an object itself if it's not a list, however if you look in the `_nested_dicts_to_dotted_keys` it eventually calls itself again if it finds that it has nested dicts. The idea is that we want all the data to be hashable because it can be useful for things like uniqueness tests, so we recursively go through the dict and convert data types as needed.